hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactionPipeline.java

```diff
@@ -0,0 +1,209 @@
+  }
+
+  public boolean pushHead(MutableSegment segment) {
+    ImmutableSegment immutableSegment = SegmentFactory.instance().
+        createImmutableSegment(segment);
+    lock.lock();
+    try {
+      return addFirst(immutableSegment);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public ImmutableSegment pullTail() {
+    lock.lock();
+    try {
+      if(pipeline.isEmpty()) {
+        return EMPTY_MEM_STORE_SEGMENT;
+      }
+      return removeLast();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public VersionedSegmentsList getVersionedList() {
+    lock.lock();
+    try {
+      LinkedList<ImmutableSegment> segmentList = new LinkedList<ImmutableSegment>(pipeline);
+      VersionedSegmentsList res = new VersionedSegmentsList(segmentList, version);
+      return res;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Swaps the versioned list at the tail of the pipeline with the new compacted segment.
+   * Swapping only if there were no changes to the suffix of the list while it was compacted.
+   * @param versionedList tail of the pipeline that was compacted
+   * @param segment new compacted segment
+   * @return true iff swapped tail with new compacted segment
+   */
+  public boolean swap(VersionedSegmentsList versionedList, ImmutableSegment segment) {
+    if(versionedList.getVersion() != version) {
+      return false;
+    }
+    lock.lock();
+    try {
+      if(versionedList.getVersion() != version) {
+        return false;
+      }
+      LinkedList<ImmutableSegment> suffix = versionedList.getStoreSegments();
+      boolean valid = validateSuffixList(suffix);
+      if(!valid) return false;
+      LOG.info("Swapping pipeline suffix with compacted item. "
+          +"Just before the swap the number of segments in pipeline is:"
+          +versionedList.getStoreSegments().size()
+          +", and the number of cells in new segment is:"+segment.getCellsCount());
+      swapSuffix(suffix,segment);
+      if(region != null) {
+        // update the global memstore size counter
+        long suffixSize = CompactingMemStore.getSegmentListSize(suffix);
+        long newSize = CompactingMemStore.getSegmentSize(segment);
+        long delta = suffixSize - newSize;
+        long globalMemstoreSize = region.addAndGetGlobalMemstoreSize(-delta);
+        LOG.info("Suffix size: "+ suffixSize+" compacted item size: "+newSize+
+            " globalMemstoreSize: "+globalMemstoreSize);
+      }
+      return true;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean isEmpty() {
+    return pipeline.isEmpty();
+  }
+
+  public LinkedList<Segment> getStoreSegmentList() {
+    lock.lock();
+    try {
+      LinkedList<Segment> res = new LinkedList<Segment>(pipeline);
+      return res;
+    } finally {
+      lock.unlock();
+    }
+
+  }
+
+  public long size() {
+    return pipeline.size();
+  }
+
+  private boolean validateSuffixList(LinkedList<ImmutableSegment> suffix) {
```

Comment:
Isn't that check a "spending time for over-protection"? The comparison of versions should include covering from all possible changes...